head/src/main/kotlin/io/qalipsis/core/head/report/ReportService.kt

```diff
@@ -60,5 +69,31 @@ internal interface ReportService {
      * @param size the maximum count of results in a page
      * @param page the 0-based index of the page to fetch
      */
-    suspend fun search(tenant: String, username: String, filters: Collection<String>, sort: String?, page: Int, size: Int): Page<Report>
+    suspend fun search(
+        tenant: String,
+        username: String,
+        filters: Collection<String>,
+        sort: String?,
+        page: Int,
+        size: Int
+    ): Page<Report>
+
+    /**
+     * Builds and renders a report in pdf form.
+     *
+     * @param tenant the reference of the tenant owning the report
+     * @param creator username of the currently authenticated user
+     * @param reference the unique identifier of the report
+     */
+    suspend fun render(tenant: String, creator: String, reference: String): String
+
+
+    /**
+     * Returns the report as ByteArray.
+     *
+     * @param tenant the reference of the tenant owning the report
+     * @param username the username of the currently authenticated user
+     * @param taskReference the unique identifier of the task data to be downloaded
+     */
+    suspend fun download(tenant: String, username: String, taskReference: String): ByteArray
```

Comment:
I already have this @Produces(MediaType.APPLICATION_OCTET_STREAM) on the download controller. Doesn't it suffice?

Reply:
On one side, you are right. The annotation should be on the controller, not the service. I misread.
However, the documentation of the service does not let us know what is returned. Furthermore, I would prefer that the service returns a class containing the binary payload and the name of the file, as created by the service, to avoid having always the same name.